### PR TITLE
Fix assertions and improve error messages when adding submobjects

### LINKED ISSUE #3755: `VGroup` cannot have a `Mobject` as a child, but `VMobject` can
## Description of bug / unexpected behavior
`Mobject` methods like `add()` assert that all the values passed are also instances of `Mobject`, before adding them as children (into `Mobject.submobjects`).

`VGroup` overrides `add()` to go even further, by asserting that the values are instances of the `VMobject` subclass, being more restrictive than the original assertion.

However, **the `VMobject` class itself does NOT have that assertion**, so you can actually add a `Mobject` to it which is not a `VMobject`.

Therefore, this is legal:
```py
Circle().add(Mobject())
```
but this raises a `TypeError`:
```py
VGroup().add(Mobject())
```

A direct consequence of this is that this is allowed:
```py
circle = Circle()
circle.add(Mobject())
group = VGroup()
group.add(circle)
```
creating a `VGroup` which contains a `Mobject` in its family, even though it's not a direct child / submobject.

## Expected behavior
The `VMobject` itself should assert that only `VMobject`s can be added to it. Adding a `Mobject` to a `VMobject` such as `Circle` should be illegal and also throw a `TypeError`. By doing this, `VGroup` would automatically do the assertion as well, as it's a subclass of `VMobject`.

### PR DESCRIPTION
**Pull request by chopan050**
Closes #3755.

Initially this PR only improved 2 things:
1. When adding into a `Mobject` a value which is not a `Mobject`, or there's an attempt to add a mobject to itself, **the error messages are now more descriptive**, indicating:
    - what's the value being added
    - what's its type
    - at which index is the issue happening
    - into which Mobject you're attempting to add the value

    Example:
```py
>>> img = ImageMobject("cat.jpg")
>>> img.add(Circle(), 3)
# TypeError: Only values of type Mobject can be added as submobjects of ImageMobject, but the value 3 (at index 1) is of type int.
>>> img.add(img)
# ValueError: Cannot add ImageMobject as a submobject of itself (at index 0).
```
2. **Refactor the code by creating a `Mobject._assert_valid_submobjects()` method which makes this check, and calling it on every method which required this assertion, which are `Mobject.add()`, `Mobject.insert()` and `Mobject.add_to_back()`.** Previously, those methods duplicated the code for checking the submobjects, but in different ways, with very different error messages, and occasionally not always checking the types of the submobjects.

But I noticed that, while `VGroup` overrided the `add()` method to be more restrictive (by checking that the submobjects are all instances of `VMobject`), the `VMobject` class itself didn't do that, which causes the issue #3755: you could add a `Mobject` to a `VMobject` such as `Circle`, and a `Mobject` could be part of the family of a `VGroup` without being a direct child of it.

Plus, the other methods `insert()` and `add_to_back()` weren't overridden, so one could still do this:
```py
>>> VGroup().add_to_back(Mobject())
VGroup(Mobject)
```

So, this PR also fixes that.

3. **Any `VMobject`, not only `VGroup`s, now asserts on `add()`, `insert()` and `add_to_back()` that all the submobjects are instances of `VMobject`.** To do that, it suffices to override the `_assert_valid_submobjects()` method in `VMobject`. It is not necessary to actually touch the `add()`, `insert()` and `add_to_back()` methods. _Because the `VGroup.add()` docstrings have an useful example, I decided to leave that override in the code, even if it now contains only a `return super().add(*vmobjects)` line._

```py
>>> Circle().add(Mobject())
# TypeError: Only values of type VMobject can be added as submobjects of Circle, but the value Mobject (at index 0) is of type Mobject. You can try adding this value into a Group instead.
```

Finally:

4. Repeat all the previous steps for `OpenGLMobject` and `OpenGLVMobject` to avoid breaking tests, and for more consistency.
5. Update the tests to take into account the more descriptive error messages.

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

File: tests/module/mobject/mobject/test_mobject.py [hunk 1]
```diff
@@ -29,12 +29,20 @@ def test_mobject_add():
     obj = Mobject()
 
     # a Mobject cannot contain itself
-    with pytest.raises(ValueError):
-        obj.add(obj)
+    with pytest.raises(ValueError) as add_self_info:
+        obj.add(Mobject(), obj, Mobject())
+    print(str(add_self_info.value))
```
Comment (JasonGrace2282) [inline tests/module/mobject/mobject/test_mobject.py:34]:
```suggestion
```
Is there a need to print it?

**Review by JasonGrace2282 — CHANGES_REQUESTED** [review]
Changes look good to me, but I think you missed some places on the opengl side?
https://github.com/ManimCommunity/manim/blob/main/manim/mobject/opengl/opengl_vectorized_mobject.py#L1662-L1663.patch
https://github.com/ManimCommunity/manim/blob/main/manim/mobject/opengl/opengl_vectorized_mobject.py#L1776-L1777.patch

File: tests/module/mobject/mobject/test_mobject.py (hunk 1 above)
Comment (chopan050) [reply tests/module/mobject/mobject/test_mobject.py]: Whoops, that shouldn't have been there. I just removed it.

**Comment by chopan050** [comment]
Thanks for highlighting those places! I fixed them, as well as some other places I missed.

**Review by JasonGrace2282 — APPROVED** [review]

**Pull request merged**